File: CS/Sample/SortableCheckedListBox.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraEditors;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;

namespace Sample
{
    class CustomCheckedlistBox : CheckedListBoxControl
    {
        int dragSourceIndex, dragTargetIndex;
        bool isDraging;
        bool[] allowMove;

        Rectangle dragBeginRect;
        public CustomCheckedlistBox()
            : base()
        {
            SelectionMode = SelectionMode.None;
            TabStop = false;
            CheckOnClick = true;
            isDraging = false;            
        }
        protected virtual internal void SetAllowMoving(int index, bool value)
        {            
            allowMove[index] = value;
        }
        protected virtual internal void CreateAllowMovingArray() { allowMove = new bool[Items.Count]; }
        protected virtual bool IsDraging { get { return isDraging; } set { if (isDraging != value) isDraging = value; } }
        protected override BaseStyleControlViewInfo CreateViewInfo() { return new CustomCheckedListBoxViewInfo(this); }
        protected override BaseControlPainter CreatePainter() { return new CustomPainterCheckedListBox(); }
        protected new virtual CustomCheckedListBoxViewInfo ViewInfo { get { return base.ViewInfo as CustomCheckedListBoxViewInfo; } }
        protected override void OnMouseDown(System.Windows.Forms.MouseEventArgs e)
        {
            base.OnMouseDown(e);
            if (e.Button == MouseButtons.Left)
            {
                int index = this.IndexFromPoint(e.Location);
                if (index >= 0 && index < Items.Count && !allowMove[index]) return;
                IsDraging = false;
                dragSourceIndex = index;
                dragTargetIndex = dragSourceIndex;
                if (dragSourceIndex != -1)
                {
                    Size dragSize = SystemInformation.DragSize;
                    dragBeginRect = new Rectangle(new Point(e.X - (dragSize.Width / 2), e.Y - (dragSize.Height / 2)), dragSize);
                }
                else
                    dragBeginRect = Rectangle.Empty;
            }
        }
        protected override void OnMouseUp(MouseEventArgs e)
        {
            if (!IsDraging || dragBeginRect == Rectangle.Empty) base.OnMouseUp(e);
            if (dragBeginRect == Rectangle.Empty) return;
            if (dragSourceIndex != -1 && dragTargetIndex != dragSourceIndex)
                ChangeItemsPositionCore(dragSourceIndex, dragTargetIndex);

            dragBeginRect = Rectangle.Empty;
        }
        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);
            if ((e.Button & MouseButtons.Left) == MouseButtons.Left)
                if (dragBeginRect != Rectangle.Empty && !dragBeginRect.Contains(e.X, e.Y))
                {
                    IsDraging = true;
                    dragTargetIndex = this.IndexFromPoint(e.Location);
                    if (dragTargetIndex == -1)
                        if (e.Y < this.ViewInfo.GetItemRectangle(0).Bottom) dragTargetIndex = 0;
                    CheckedListBoxViewInfo.CheckedItemInfo info = ViewInfo.GetItemByIndex(dragSourceIndex);
                    if (info != null) info.PaintAppearance.ForeColor = Color.Red;
                    ViewInfo.MarkItem(dragTargetIndex, dragSourceIndex);
                }
        }
        protected virtual void ChangeItemsPositionCore(int source, int target)
        {
            CorrectAllowMove(source, target);
            if (target == -1)
            {
                Items.Add(Items[source]);
            }
            else
            {
                Items.Insert(target, Items[source]);
                if (source > target) source++;
            }
            Items.RemoveAt(source);         
        }

        protected virtual void CorrectAllowMove(int source, int target)
        {
            bool b = allowMove[source];
            if (target == -1) target = Items.Count-1;
            if (target > source)
                for (int i = source; i < target; i++) allowMove[i] = allowMove[i+1];
            else
                for (int i = source; i > target; i--) allowMove[i] = allowMove[i-1];
            allowMove[target] = b;
        }
    }
    public class CustomCheckedListBoxViewInfo : CheckedListBoxViewInfo
    {
        Color dragDropLineColor;
        public CustomCheckedListBoxViewInfo(CheckedListBoxControl listBox) : base(listBox) { }
        protected override ItemInfo CreateItemInfo(Rectangle bounds, object item, string text, int index)
        {
            CheckedItemInfo info = base.CreateItemInfo(bounds, item, text, index) as CheckedItemInfo;
            CustomCheckedItemInfo patchedInfo = new CustomCheckedItemInfo(info);
            return patchedInfo;
        }
        public virtual Color DragDropLineColor { get { return dragDropLineColor; } set { if (dragDropLineColor != value) dragDropLineColor = value; } }
        protected internal virtual void UnderlineItem(int index)
        {
            CustomCheckedItemInfo info = base.GetItemByIndex(index) as CustomCheckedItemInfo;
            if (info != null)
                info.IsUnderLine = true;
        }
        protected internal virtual void OverlineItem(int index)
        {
            CustomCheckedItemInfo info = base.GetItemByIndex(index) as CustomCheckedItemInfo;
            if (info != null)
                info.IsOverLine = true;
        }
        protected virtual internal int ItemCountAccessMethod() { return ItemCount; }
        protected internal virtual void DropLine()
        {
            foreach (CustomCheckedItemInfo info in VisibleItems)                
                info.DropLine();
        }
        protected internal virtual void MarkItem(int targetIndex, int sourceIndex)
        {
            DropLine();
            dragDropLineColor = Color.Red;
            if ((targetIndex == sourceIndex) || (targetIndex == sourceIndex + 1) ||
                (sourceIndex == ItemCount  - 1 && targetIndex == -1)) dragDropLineColor = Color.LightGray;
            if (targetIndex == -1)
                UnderlineItem( ItemCount - 1);
            else
                OverlineItem(targetIndex);
            if (targetIndex > 0) 
                UnderlineItem(targetIndex - 1);
            OwnerControl.Invalidate();
        }

        public class CustomCheckedItemInfo : CheckedItemInfo
        {
            bool isUnderLine, isOverLine;
            public CustomCheckedItemInfo(Rectangle rect, object item, string text, int index, CheckState checkState, bool enabled)
                : base(rect, item, text, index, checkState, enabled)
            {
                DropLine();
            }
            public CustomCheckedItemInfo(CheckedItemInfo info)
                : this(info.Bounds, info.Item, info.Text, info.Index, info.CheckArgs.CheckState, info.Enabled)
            {
                this.CheckArgs.Assign(info.CheckArgs);
                this.TextRect = info.TextRect;
            }
            protected virtual internal bool IsUnderLine { get { return isUnderLine; } set { isUnderLine = value; } }
            protected virtual internal bool IsOverLine { get { return isOverLine; } set { isOverLine = value; } }            
            protected virtual internal void DropLine()
            {
                isUnderLine = false;
                isOverLine = false;
            }
        }
    }
    public class CustomPainterCheckedListBox : PainterCheckedListBox
    {
        const int lineWidth = 1;
        protected override void DrawItemCore(ControlGraphicsInfoArgs info, BaseListBoxViewInfo.ItemInfo itemInfo, ListBoxDrawItemEventArgs e)
        {
            base.DrawItemCore(info, itemInfo, e);
            Sample.CustomCheckedListBoxViewInfo.CustomCheckedItemInfo customInfo =
                itemInfo as Sample.CustomCheckedListBoxViewInfo.CustomCheckedItemInfo;
            if (customInfo == null) return;
            Rectangle rec = new Rectangle(itemInfo.Bounds.Location, new Size(itemInfo.Bounds.Width, lineWidth));
            Color lineColor = ((CustomCheckedListBoxViewInfo)info.ViewInfo).DragDropLineColor;
            if (customInfo.IsOverLine)
            {
                if (customInfo.Index == 0) rec.Height++;
                info.Graphics.FillRectangle(info.Cache.GetSolidBrush(lineColor), rec);
            }
            if (customInfo.IsUnderLine)
            {
                rec.Offset(0, itemInfo.Bounds.Height - lineWidth);
                if (customInfo.Index == ((CustomCheckedListBoxViewInfo)info.ViewInfo).ItemCountAccessMethod() - 1) rec.Height++;
                info.Graphics.FillRectangle(info.Cache.GetSolidBrush(lineColor), rec);
            }
        }
    }
}

File: VB/Sample/SortableCheckedListBox.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraEditors
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing

Namespace Sample
	Friend Class CustomCheckedlistBox
		Inherits CheckedListBoxControl

		Private dragSourceIndex, dragTargetIndex As Integer
'INSTANT VB NOTE: The variable isDraging was renamed since Visual Basic does not allow variables and other class members to have the same name:
		Private isDraging_Renamed As Boolean
		Private allowMove() As Boolean

		Private dragBeginRect As Rectangle
		Public Sub New()
			MyBase.New()
			SelectionMode = SelectionMode.None
			TabStop = False
			CheckOnClick = True
			isDraging_Renamed = False
		End Sub
		Protected Friend Overridable Sub SetAllowMoving(ByVal index As Integer, ByVal value As Boolean)
			allowMove(index) = value
		End Sub
		Protected Friend Overridable Sub CreateAllowMovingArray()
			allowMove = New Boolean(Items.Count - 1){}
		End Sub
		Protected Overridable Property IsDraging() As Boolean
			Get
				Return isDraging_Renamed
			End Get
			Set(ByVal value As Boolean)
				If isDraging_Renamed <> value Then
					isDraging_Renamed = value
				End If
			End Set
		End Property
		Protected Overrides Function CreateViewInfo() As BaseStyleControlViewInfo
			Return New CustomCheckedListBoxViewInfo(Me)
		End Function
		Protected Overrides Function CreatePainter() As BaseControlPainter
			Return New CustomPainterCheckedListBox()
		End Function
		Protected Shadows Overridable ReadOnly Property ViewInfo() As CustomCheckedListBoxViewInfo
			Get
				Return TryCast(MyBase.ViewInfo, CustomCheckedListBoxViewInfo)
			End Get
		End Property
		Protected Overrides Sub OnMouseDown(ByVal e As System.Windows.Forms.MouseEventArgs)
			MyBase.OnMouseDown(e)
			If e.Button = MouseButtons.Left Then
				Dim index As Integer = Me.IndexFromPoint(e.Location)
				If index >= 0 AndAlso index < Items.Count AndAlso (Not allowMove(index)) Then
					Return
				End If
				IsDraging = False
				dragSourceIndex = index
				dragTargetIndex = dragSourceIndex
				If dragSourceIndex <> -1 Then
					Dim dragSize As Size = SystemInformation.DragSize
					dragBeginRect = New Rectangle(New Point(e.X - (dragSize.Width \ 2), e.Y - (dragSize.Height \ 2)), dragSize)
				Else
					dragBeginRect = Rectangle.Empty
				End If
			End If
		End Sub
		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			If (Not IsDraging) OrElse dragBeginRect = Rectangle.Empty Then
				MyBase.OnMouseUp(e)
			End If
			If dragBeginRect = Rectangle.Empty Then
				Return
			End If
			If dragSourceIndex <> -1 AndAlso dragTargetIndex <> dragSourceIndex Then
				ChangeItemsPositionCore(dragSourceIndex, dragTargetIndex)
			End If

			dragBeginRect = Rectangle.Empty
		End Sub
		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			If (e.Button And MouseButtons.Left) = MouseButtons.Left Then
				If dragBeginRect <> Rectangle.Empty AndAlso (Not dragBeginRect.Contains(e.X, e.Y)) Then
					IsDraging = True
					dragTargetIndex = Me.IndexFromPoint(e.Location)
					If dragTargetIndex = -1 Then
						If e.Y < Me.ViewInfo.GetItemRectangle(0).Bottom Then
							dragTargetIndex = 0
						End If
					End If
					Dim info As CheckedListBoxViewInfo.CheckedItemInfo = ViewInfo.GetItemByIndex(dragSourceIndex)
					If info IsNot Nothing Then
						info.PaintAppearance.ForeColor = Color.Red
					End If
					ViewInfo.MarkItem(dragTargetIndex, dragSourceIndex)
				End If
			End If
		End Sub
		Protected Overridable Sub ChangeItemsPositionCore(ByVal source As Integer, ByVal target As Integer)
			CorrectAllowMove(source, target)
			If target = -1 Then
				Items.Add(Items(source))
			Else
				Items.Insert(target, Items(source))
				If source > target Then
					source += 1
				End If
			End If
			Items.RemoveAt(source)
		End Sub

		Protected Overridable Sub CorrectAllowMove(ByVal source As Integer, ByVal target As Integer)
			Dim b As Boolean = allowMove(source)
			If target = -1 Then
				target = Items.Count-1
			End If
			If target > source Then
				For i As Integer = source To target - 1
					allowMove(i) = allowMove(i+1)
				Next i
			Else
				For i As Integer = source To target + 1 Step -1
					allowMove(i) = allowMove(i-1)
				Next i
			End If
			allowMove(target) = b
		End Sub
	End Class
	Public Class CustomCheckedListBoxViewInfo
		Inherits CheckedListBoxViewInfo

'INSTANT VB NOTE: The variable dragDropLineColor was renamed since Visual Basic does not allow variables and other class members to have the same name:
		Private dragDropLineColor_Renamed As Color
		Public Sub New(ByVal listBox As CheckedListBoxControl)
			MyBase.New(listBox)
		End Sub
		Protected Overrides Function CreateItemInfo(ByVal bounds As Rectangle, ByVal item As Object, ByVal text As String, ByVal index As Integer) As ItemInfo
			Dim info As CheckedItemInfo = TryCast(MyBase.CreateItemInfo(bounds, item, text, index), CheckedItemInfo)
			Dim patchedInfo As New CustomCheckedItemInfo(info)
			Return patchedInfo
		End Function
		Public Overridable Property DragDropLineColor() As Color
			Get
				Return dragDropLineColor_Renamed
			End Get
			Set(ByVal value As Color)
				If dragDropLineColor_Renamed <> value Then
					dragDropLineColor_Renamed = value
				End If
			End Set
		End Property
		Protected Friend Overridable Sub UnderlineItem(ByVal index As Integer)
			Dim info As CustomCheckedItemInfo = TryCast(MyBase.GetItemByIndex(index), CustomCheckedItemInfo)
			If info IsNot Nothing Then
				info.IsUnderLine = True
			End If
		End Sub
		Protected Friend Overridable Sub OverlineItem(ByVal index As Integer)
			Dim info As CustomCheckedItemInfo = TryCast(MyBase.GetItemByIndex(index), CustomCheckedItemInfo)
			If info IsNot Nothing Then
				info.IsOverLine = True
			End If
		End Sub
		Protected Friend Overridable Function ItemCountAccessMethod() As Integer
			Return ItemCount
		End Function
		Protected Friend Overridable Sub DropLine()
			For Each info As CustomCheckedItemInfo In VisibleItems
				info.DropLine()
			Next info
		End Sub
		Protected Friend Overridable Sub MarkItem(ByVal targetIndex As Integer, ByVal sourceIndex As Integer)
			DropLine()
			dragDropLineColor_Renamed = Color.Red
			If (targetIndex = sourceIndex) OrElse (targetIndex = sourceIndex + 1) OrElse (sourceIndex = ItemCount - 1 AndAlso targetIndex = -1) Then
				dragDropLineColor_Renamed = Color.LightGray
			End If
			If targetIndex = -1 Then
				UnderlineItem(ItemCount - 1)
			Else
				OverlineItem(targetIndex)
			End If
			If targetIndex > 0 Then
				UnderlineItem(targetIndex - 1)
			End If
			OwnerControl.Invalidate()
		End Sub

		Public Class CustomCheckedItemInfo
			Inherits CheckedItemInfo

'INSTANT VB NOTE: The variable isUnderLine was renamed since Visual Basic does not allow variables and other class members to have the same name:
'INSTANT VB NOTE: The variable isOverLine was renamed since Visual Basic does not allow variables and other class members to have the same name:
			Private isUnderLine_Renamed, isOverLine_Renamed As Boolean

			Public Sub New(ByVal ownerControl As BaseListBoxControl, ByVal rect As Rectangle, ByVal item As Object, ByVal text As String, ByVal index As Integer, ByVal checkState As CheckState, ByVal enabled As Boolean)
				MyBase.New(ownerControl, rect, item, text, index, checkState, enabled)
				DropLine()
			End Sub
			Public Sub New(ByVal info As CheckedItemInfo)
				Me.New(info.ListBoxControl, info.Bounds, info.Item, info.Text, info.Index, info.CheckArgs.CheckState, info.Enabled)
				Me.CheckArgs.Assign(info.CheckArgs)
				Me.TextRect = info.TextRect
			End Sub
			Protected Friend Overridable Property IsUnderLine() As Boolean
				Get
					Return isUnderLine_Renamed
				End Get
				Set(ByVal value As Boolean)
					isUnderLine_Renamed = value
				End Set
			End Property
			Protected Friend Overridable Property IsOverLine() As Boolean
				Get
					Return isOverLine_Renamed
				End Get
				Set(ByVal value As Boolean)
					isOverLine_Renamed = value
				End Set
			End Property
			Protected Friend Overridable Sub DropLine()
				isUnderLine_Renamed = False
				isOverLine_Renamed = False
			End Sub
		End Class
	End Class
	Public Class CustomPainterCheckedListBox
		Inherits PainterCheckedListBox

		Private Const lineWidth As Integer = 1
		Protected Overrides Sub DrawItemCore(ByVal info As ControlGraphicsInfoArgs, ByVal itemInfo As BaseListBoxViewInfo.ItemInfo, ByVal e As ListBoxDrawItemEventArgs)
			MyBase.DrawItemCore(info, itemInfo, e)
			Dim customInfo As Sample.CustomCheckedListBoxViewInfo.CustomCheckedItemInfo = TryCast(itemInfo, Sample.CustomCheckedListBoxViewInfo.CustomCheckedItemInfo)
			If customInfo Is Nothing Then
				Return
			End If
			Dim rec As New Rectangle(itemInfo.Bounds.Location, New Size(itemInfo.Bounds.Width, lineWidth))
			Dim lineColor As Color = CType(info.ViewInfo, CustomCheckedListBoxViewInfo).DragDropLineColor
			If customInfo.IsOverLine Then
				If customInfo.Index = 0 Then
					rec.Height += 1
				End If
				e.Cache.FillRectangle(lineColor, rec)
			End If
			If customInfo.IsUnderLine Then
				rec.Offset(0, itemInfo.Bounds.Height - lineWidth)
				If customInfo.Index = CType(info.ViewInfo, CustomCheckedListBoxViewInfo).ItemCountAccessMethod() - 1 Then
					rec.Height += 1
				End If
				e.Cache.FillRectangle(lineColor, rec)
			End If
		End Sub
	End Class
End Namespace
